content/js1/sprints/2/prep/index.md

```diff
@@ -8,4 +8,406 @@ backlog= 'Module-JS1'
+```
+
+### 💻 Playing computer
+
+To understand how our function currently behaves we must start building an understanding of how our code is executed. We might currently have some ideas of how code is executed in our program:
+
+```mermaid
+mindmap
+Root[How is a piece of code executed?]
+    B[Lines of code are executed from top to bottom]
+    C[Variable declarations mean data is stored in memory]
+    D[Function call means running the instructions inside the function]
+    E[When returning from a function we get an output]
+```
+
+### ➡️ Current line of code
+
+We can make use of execution diagrams to visualise how a program is executed by a computer.
+
+Firstly, we need a way to track the **line of code that is currently executing.** In the diagrams below, we use an arrow to indicate which line of code is currently being executed. Each line is an instruction to the computer that will produce some effect. For example, a line of code with a variable declaration means "store a new variable with a particular value".
+
+Let's consider an example:
+
+<iframe width="800" height="300" frameborder="0" src="https://pythontutor.com/iframe-embed.html#code=const%20name%20%3D%20'Gayle'%3B&codeDivHeight=400&codeDivWidth=350&cumulative=false&curInstr=0&heapPrimitives=nevernest&origin=opt-frontend.js&py=js&rawInputLstJSON=%5B%5D&textReferences=false"> </iframe>
+
+In the example above, we have a single variable declaration:
+
+```js
+const name = "Gayle";
+```
+
+which means store a string `"Gayle"` in memory.
+
+### 🖼️ Storing things in memory
+
+As variables are declared, we need a way to keep track of the memory that is stored as our program executes.
+In our diagrams, we keep track of memory using a _frame_.
+
+For now, we can think of a frame as a space in our diagram that represents the variables stored in memory in some part of our code.
+
+The _global frame_ is always the first frame that gets created when our program starts executing. We can think of the _global frame_ as a location for variables created outside of any function. When we run the variable declaration above, we get a string `"Gayle"` stored in the Global frame:
+
+![global-frame](global-frame.png)
+
+### 🏃 Executing functions
+
+So far we've seen 2 things:
+
+- lines of code are executed from top to bottom
+- we keep track of variables stored in memory using a global frame
+
+However, we need to consider what happens when we declare and call a function in our program.
+Let's consider another example:
+
+<iframe width="850" height="375" frameborder="0" src="https://pythontutor.com/iframe-embed.html#code=const%20firstName%20%3D%20'German'%3B%0Aconst%20lastName%20%3D%20'Bencii'%3B%0A%0Afunction%20makeGreeting%28%29%20%7B%0A%20%20%20%20const%20greeting%20%3D%20%60Hello%20there%60%3B%20%0A%7D%0A%0AmakeGreeting%28%29%3B&codeDivHeight=400&codeDivWidth=350&cumulative=false&curInstr=5&heapPrimitives=nevernest&origin=opt-frontend.js&py=js&rawInputLstJSON=%5B%5D&textReferences=false"> </iframe>
```

Comment:
In this example in the iFrame, the first thing isn't actually the function declaration, it's the `firstName` variable. I think we should edit the example so it actually _does_ start with the function declaration (we can come back at some later time to explain "Surprise! Functions can actually be defined out of order!"), but I think that's too much noise for still understanding the idea of scope at all.

Reply:
Yeah the example below was correct - hoisting is noise at this point.